File: Jump_game_BFS.py
# BFS
# Time complexity : O(n^k)
# Space complexity : O(n^k)
# Leetcode : Time Limited exceeded

from collections import deque
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        # if the len of the list is less than 2, return True
        if len(nums) < 2:
            return True
        
        # find the length of the list, make a queue and append the first index to the queue
        n = len(nums)
        q = deque([0])
        
        # traverse until the queue is empty
        while q:
            # pop the first index
            i = q.popleft()
            
            # traverse over all the numbers from 1 till the end
            for j in range(1,nums[i]+1):
                # add the index
                idx = i + j
                
                # check if we have reached the end of the list, if so then return True else we append the index to the queue
                if idx >= n-1:
                    return True
                q.append(idx)
        
        return False

File: Jump-Game-BFS.java
// Time Complexity : O(N)

// Space Complexity : O(2N)

// Did this code successfully run on Leetcode : YES

// Appoarch: Using BFS, push the 0th node in the queue and the set. start exploring the children of 
// 0th node and add to queue. If the new idx to the last idx, return true. If not,
// check if its present in the set and add to set and queue.

// 55. Jump Game

import java.util.HashSet;
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public boolean canJump(int[] nums) {
        if(nums.length == 0 || nums.length == 1) return true;
        HashSet<Integer> set = new HashSet<>();
        Queue<Integer> q = new LinkedList<>();
        q.add(0);
        set.add(0);
        while(!q.isEmpty()){
            int curr = q.poll();
            for(int i=1; i <= nums[curr]; i++){
                int newIdx = curr + i;
                if(newIdx == nums.length-1) return true;
                if(!set.contains(newIdx)){
                    set.add(newIdx);
                    q.add(newIdx);
                }
            }
        }
        return false;
    }
}